File: 백준/Gold/14500. 테트로미노/테트로미노.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, a[500][500], visited[500][500], ans, dy[] = { 0,0,1,-1 }, dx[] = { 1,-1,0,0 };

void dfs(int y, int x, int cnt, int sum) {
	if (cnt == 4) {
		ans = max(ans, sum);
		return;
	}

	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny >= 0 && nx >= 0 && ny < n && nx < m && !visited[ny][nx]) {
			visited[ny][nx] = 1;
			dfs(ny, nx, cnt + 1, sum + a[ny][nx]);
			visited[ny][nx] = 0;
		}
	}
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++)
			cin >> a[i][j];
	}

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			visited[i][j] = 1;
			dfs(i, j, 1, a[i][j]);
			visited[i][j] = 0;
			
			if (i - 1 >= 0 && j + 2 < m) 
				ans = max(ans, a[i][j] + a[i][j + 1] + a[i][j + 2] + a[i - 1][j + 1]);
			if (i + 1 < n && j + 2 < m)
				ans = max(ans, a[i][j] + a[i][j + 1] + a[i][j + 2] + a[i + 1][j + 1]);
			if (i + 2 < n && j + 1 < m)
				ans = max(ans, a[i][j] + a[i + 1][j] + a[i + 2][j] + a[i + 1][j + 1]);
			if (i + 2 < n && j - 1 >= 0)
				ans = max(ans, a[i][j] + a[i + 1][j] + a[i + 2][j] + a[i + 1][j - 1]);
		}
	}

	cout << ans;

	return 0;
}

File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m, ans;
	static int[][] a;
	static boolean visited[][];
	static int[] dy = { -1, 0, 1, 0 };
	static int[] dx = { 0, 1, 0, -1 };

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		a = new int[n][m];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < m; j++) {
				a[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		visited = new boolean[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				visited[i][j] = true;
				dfs(i, j, a[i][j], 1);
				visited[i][j] = false;

				if (i - 1 >= 0 && i + 1 < n && j + 1 < m) {
					ans = Math.max(ans, a[i][j] + a[i - 1][j] + a[i + 1][j] + a[i][j + 1]);
				}
				if (i + 1 < n && j - 1 >= 0 && j + 1 < m) {
					ans = Math.max(ans, a[i][j] + a[i + 1][j] + a[i][j - 1] + a[i][j + 1]);
				}
				if (i - 1 >= 0 && j - 1 >= 0 && j + 1 < m) {
					ans = Math.max(ans, a[i][j] + a[i - 1][j] + a[i][j - 1] + a[i][j + 1]);
				}
				if (i - 1 >= 0 && i + 1 < n && j - 1 >= 0) {
					ans = Math.max(ans, a[i][j] + a[i - 1][j] + a[i + 1][j] + a[i][j - 1]);
				}
			}
		}

		bw.write(ans + "");
		bw.flush();
	}

	static void dfs(int y, int x, int sum, int depth) {
		if (depth == 4) {
			ans = Math.max(ans, sum);
			return;
		}

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];
			if (ny >= n || nx >= m || ny < 0 || nx < 0 || visited[ny][nx]) {
				continue;
			}

			visited[ny][nx] = true;
			dfs(ny, nx, sum + a[ny][nx], depth + 1);
			visited[ny][nx] = false;
		}
	}
}
